platform/linux-generic/include/odp_ring_st_internal.h

```diff
@@ -0,0 +1,118 @@
+/* Copyright (c) 2018, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#ifndef ODP_RING_ST_INTERNAL_H_
+#define ODP_RING_ST_INTERNAL_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <odp/api/hints.h>
+#include <odp_align_internal.h>
+
+/* Basic ring for single thread usage. Operations must be synchronized by using
+ * locks (or other means), when multiple threads use the same ring. */
+typedef struct {
+	uint32_t head;
+	uint32_t tail;
+	uint32_t mask;
+	uint32_t *data;
+
+} ring_st_t;
+
+/* Initialize ring. Ring size must be a power of two. */
+static inline void ring_st_init(ring_st_t *ring, uint32_t *data, uint32_t size)
+{
+	ring->head = 0;
+	ring->tail = 0;
+	ring->mask = size - 1;
+	ring->data = data;
+}
+
+/* Dequeue data from the ring head. Max_num is smaller than ring size.*/
+static inline uint32_t ring_st_deq_multi(ring_st_t *ring, uint32_t data[],
+					 uint32_t max_num)
+{
+	uint32_t head, tail, mask, idx;
+	uint32_t num, i;
+
+	head = ring->head;
+	tail = ring->tail;
+	mask = ring->mask;
+	num  = tail - head;
```

Comment:
There's no need for abs(). Since it's all uint32_t variables, wrap a round is handled already.
An example in 8bits:
0xff - 0xfd = 0x02
0x00 - 0xfe = 0x02
0x01 - 0xff = 0x02
0x02 - 0x00 = 0x02

This passes both gcc and clang, and is used already in the other ring implementation see ring_deq_multi().

Reply:
The point is that ring size will never be close to 4 billion entries. E.g. currently tail is always max 4096 larger than head. Your example above is based assumption of 4 billion entry ring. Overflow is avoided when ring size if <2 billion, as 32 bit indexes can be still used to calculate number of items correctly.

Reply:
Not true. See the comment below where it's clear that `head` and `tail` are free running and thus will eventually wrap as described above. If these were `uint64_t` variables you could argue that the wraps would take centuries and hence are nothing to worry about, but being `uint32_t` variables this should be expected to happen regularly on heavily used queues. 

`abs()` adds no overhead since compilers treat it as an intrinsic.

Reply:
It's unsigned 32 bit arithmetic, not signed integer arithmetic. Also, tail is always ahead of head, but no more than queue size (e.g. 4k), So, e.g. when tail is 0, and head is 0xffffffff the number of items in ring is 0 - 0xffffffff = 1

```
head            tail
hex    [index], hex    [index], num, num_free
0xfffffffe [6], 0xffffffff [7], 1,   7
0xffffffff [7], 0x00000000 [0], 1,   7
0x00000000 [0], 0x00000001 [1], 1,   7
```

Reply:
OK